src/util/parse_options.cpp

```diff
@@ -22,13 +22,27 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "cmdline.h"
 #include "exception_utils.h"
 #include "exit_codes.h"
+#include "message.h"
 #include "signal_catcher.h"
 
 parse_options_baset::parse_options_baset(
-  const std::string &_optstring, int argc, const char **argv)
+  const std::string &_optstring, int argc, const char **argv, messaget &l)
+  : log(l)
 {
   std::string optstring=std::string("?h(help)")+_optstring;
   parse_result=cmdline.parse(argc, argv, optstring.c_str());
+
+  // DO NOT USE l HERE!
+  //
+  // The usual pattern of use is that the application class inherits from
+  // messaget and parse_options_baset using a member variable of type
+  // message_handlert to construct the messaget part.
+  //
+  // C++'s rules of initialisation mean that the constructors for
+  // messaget and then parse_options_base run before those of message_handlert.
+  // This means that the messaget object has a reference to a
+  // an uninitialised object.  Using it here will likely cause a
+  // hard to debug failure somewhere in the messaget code.
```

Comment:
Would all that actually be pretty safe if we were using the `message_handlert` instead of the `messaget`? There arguably still is a risk that the message handler is not properly configured just yet and output would not end up formatted as desired, but it should be safe.

Reply:
The `message_handlert` won't be because it is a member variable.  This is an issue with the current "inherit from `messaget`" approach.